### ISSUE TITLE
Cannot run angular 2+ from file:/// - looks like 'base href="/"' is the issue

### ISSUE THREAD

**Issue by jacodv** [angular/angular#13948]
**I'm submitting a ...**  (check one with "x")
```
[x ] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

Windows development PC
ng new test-local
cd test-local
ng serve
See content of index.html and no errors in the console in chrome on http://localhost:4200

ng build
Navigate to /dist and open index.html in Chrome
Do not see content of page and 404 errors in console in chrome on file:///fullpath/index.html

**Expected behavior**

Expected to see the same result on http:// and file:/// views.  This is currently the case on angular 1.6.x

**Minimal reproduction of the problem with instructions**

see current behaviour to reproduce

**What is the motivation / use case for changing the behavior?**

We have many applications that serve up SPA websites from file.  iOS and OSX do not allow the hosting of a web server in your application.

**Please tell us about your environment:**

Windows 10, Sublime 3

* **Angular version:** 2.0.X

Latest as of today - 16 January 2017 - ^2.1.0

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

All
 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
TypeScript ~2.0.3

* **Node (for AoT issues):** `node --version` =   
node -v : 6.9.1
npm -v: 3.10.9

**Comment by DzmitryShylovich** [comment]
Pls provide a small demo on Plunkr

**Comment by jacodv** [comment]
Attached is the source.  It does not make sense to run it in plunker, because that executes the http:// protocol.  The issues lie with the file:/// protocol.  
[dist.zip](https://github.com/angular/angular/files/708489/dist.zip)

On a windows machine:
unzip attachment to "c:\temp\dist\"
Open chrome and paste: "file:///C:/temp/dist/index.html"

**Comment by DzmitryShylovich** [comment]
Sounds like a server (webpack dev server in case of cli) issue rather than angular tbh

**Comment by jacodv** [comment]
No it is not.  The issue is that the code must run without a web server like IIS / Node etc.  It should be able to run straight from the file system without being hosted on a web server.

**Comment by wKoza** [comment]
> The issue is that the code must run without a web server like IIS / Node etc

why ?

**Comment by DzmitryShylovich** [comment]
https://github.com/angular/angular-cli/issues/1231

**Comment by DzmitryShylovich** [comment]
It's a bundling issue. As you can see from the link above they will not support it in cli, but I believe you can accomplish it with a custom build. Just need to rewrite this line:
```
<script type="text/javascript" src="inline.bundle.js"></script><script type="text/javascript" src="styles.bundle.js"></script><script type="text/javascript" src="main.bundle.js"></script>
```

**Comment by jacodv** [comment]
@wKoza, We ship losts of SPA web sites that we serve up from file system in an offline mode for desktop and mobile apps.  Thus we need the file:/// option to serve the file.  Currently we use angular 1.6.x and it works.

@DzmitryShylovich, 
I will use samples from angular.io to test your theory, they do not use cli.  I will post my findings here. 

Bare with me for a second.  Look at the attached image, it shows the error and why the file not found error is thrown.  The browser is looking for the script in the root of the drive.  My guess is that this is because we have the new `<base href="/">` attribute in the html.  The folder, as per the address bar, contains all the required files.

![image](https://cloud.githubusercontent.com/assets/5814127/21989156/8bb9b16e-dc12-11e6-9106-b91b10fc4bdc.png)

**Comment by DzmitryShylovich** [comment]
> My guess is that this is because we have the new <base href="/">

you are right. Works fine without it. You can remove it and and provide base ref programatically

**Comment by wKoza** [comment]
With angular-cli, you can change base url with  ng build --base-href /xxx/zzz

**Comment by jacodv** [comment]
I cannot set the base URL at build time.  The base URL will be different from desktop to mobile app and from OS to OS.  The content (sources) should load relative from the main page.  All packages shipped to our customers are build once and consumed many times from many devices.

**Comment by jacodv** [comment]
I have tried ./ and ~/.  Even though ./ removes the errors from the console. It still does not run the angular code.  This is where you could be correct about the bundling, it might be paths in the bundled files.  But I would have thought that some sort of error will be displayed/thrown

**Comment by thelgevold** [comment]
Browsers will for security reasons disallow serving ajax requests from the file system. 
It typically gives you a `same origin policy error`. Right now I see 404 errors in your error description, but my theory is that you will see same origin policy errors if you specify base as `<base href=".">`

Generally you can load script tags from the file system, so that explains why angular 1.x works since the whole framework is downloaded as a single script tag. Ajax requests though will cause problems. This includes routing, or ajax request in general.

Basically Angular 2 is not designed to work off of the file system without a web server.
If you really need to serve an angular 2 app from the file system I would consider electron as a better solution. Otherwise you can use lite-server a very lightweight development web server.

**Comment by jacodv** [comment]
Thank you for the comprehensive response.  You might have hit the nail on the head.  I am not that familiar with electron, but from what I can see is that it is a desktop app framework.  

We need to package our mini SPA site in a single file, then it gets consumed by many platforms.  The consumers are websites (where angular 2 works), offline desktops and offline mobile apps (where angular 2 doesn't work).  My guess is that electron will not work in mobile apps and from web sites causing the single package not to work on all consumers.

I guess we are stuck with angular 1.x.x or ditching angular and replacing it with another framework.  This is very sad because we invested a lot into angular from a CI point of view.

**Comment by jacodv** [comment]
@thelgevold - Your opinion: Should I leave this open and hope that one of the **Angular Core** developers assist and add possible workarounds?  Or should I close this chapter?  I really want to move over to Angular 2, and this prevents us.  

In our Angular 1.x code we transpiled all the HTML views as javascript templates in the gulp build tasks.  That eliminated the ajax requests.  Could there be a similar workaround for us on angular 2?  I have all my templates in the .ts files, not .html files currently, and that didn't work either.

**Comment by wKoza** [comment]
@thelgevold says : 

> Basically Angular 2 is not designed to work off of the file system without a web server.

So, how to works ionic 2 for exemple ?

@jacodv , if you create an application sample with angular-cli and some call ajax via api github, it's work.

**Comment by jacodv** [comment]
@wKoza, I am sure that will; because the API is hosted.  I need a solution for the fact that angular to is not working when all the content is offline on the file system, let say 3 files [index.html, bundle.js, bundle.css].  the index page has a script reference to both the js and cs file.  All are extracted to c:\temp\test.  And you open the index file in Chrome via file:///c:/temp/test/index.html.

This does not work on angular 2 but does on angular 1.  And as per @thelgevold explanation, we believe that angular 2 load some resources via ajax / xhr request and angular 1 didn't.

**Comment by thelgevold** [comment]
@jacodv If you have everything bundled into a single file (angular + application code) with no external templates, or other dependencies, I suspect it's no longer an ajax request issue. Especially if you don't see any ajax requests being made in the network tab.

**Comment by jacodv** [comment]
@thelgevold - Now we are full circle.  I have uploaded the same code right at the beginning of this thread.  But can definitely upload it again or give more samples if required.  I have everything bundled with no dependencies.  Originally I thought it was the <base href="/"/> tag that causes it to fail.  I have tried the various options of the base URL, in some cases all errors are gone in the console, but the angular code still does not run.

**Comment by Markovy** [comment]
Plnkr examples have:
```javascript
<script>document.write('<base href="' + document.location + '" />');</script>
```
it works for me (when serve from file:///)

**Comment by jacodv** [comment]
@Markovy -  If this works then you are a legend!

**Comment by jacodv** [comment]
I have tried the HTML below.  No success.... There are no errors in the console.  Just Loading...., and this means that the angular did not work

![image](https://cloud.githubusercontent.com/assets/5814127/22117184/b167415e-de7b-11e6-97b2-abc63815c521.png)

**Comment by jacodv** [comment]
Also tried this
![image](https://cloud.githubusercontent.com/assets/5814127/22117551/1db748d0-de7d-11e6-9e94-df97b19b48e6.png)

**Comment by jacodv** [comment]
@Markovy - I am going to start from scratch and build you suggestion into the process.  I will revert with my finding tomorrow morning when I get to the office.

**Comment by jacodv** [comment]
I had to make it work - IT DOES! - @Markovy you have made my day!!!!!!

**Comment by Wancieho** [comment]
@jacodv thanks I will check it out when I get around to it but we had to find another framework as we were under time constraints :) I see you're a fellow South African.

**Comment by jacodv** [comment]
I am, based in Pta.  Which framework did you end up using?

**Comment by Wancieho** [comment]
Ye I also work in PTA. Is there a way to take this offline?

Landed up building (and still evolving) my own basic jQuery "framework" based on the principals of A2+TS.

**Comment by jacodv** [comment]
send me an email on jacodv.at.infoslips.co.za

**Comment by Koshmaar** [comment]
@jacodv Hey what did you exactly do to make this work? I'm having similar issue with very simple app that doesn't want to run from local filesystem.

By default it has  `<base href="/">` and links to scripts like `<script src="/js/shims.js?1485861650332"></script>`  Obviously the scripts are not found and nothing loads.

I can make the files be found simply by prefixing with dot so that paths are relative to curr dir `<base href="./">` and `<script src="./js/shims.js?1485861650332"></script>` - I see some code being run (navigation structure of site is created), but not "business" logic .  

In dev tools I see 

> EXCEPTION: Uncaught (in promise): SecurityError: Failed to execute 'replaceState' on 'History': A history state object with URL 'file:///E:/' cannot be created in a document with origin 'null' and URL 'file:///E:/WWW/WebdevHomepage/admin2/dist/prod/index.html'.

**Comment by jacodv** [comment]
@Koshmaar, I have reverted to all in a single folder.  Thus all scripts and other resources linked via href / src attributes are in the same folder as index.html.  Then all you need todo is omit any paths to the scripts / resources and only reference the name.

**Comment by housseindjirdeh** [comment]
@Koshmaar I'm experiencing the same thing at the moment and I also have everything in the same folder, so I'm not sure why I'm still seeing the error 🤔 

Just wanted to ask if there was anything else you did to fix that error?

**Comment by jacodv** [comment]
Nope that was all.  Let me see if I can get my test source and upload it for you to see

**Comment by Koshmaar** [comment]
I don't remember if I got this working, and after all didn't need to - I'm running app on localhost instead from file.

**Comment by jacodv** [comment]
This is a [Working Sample](https://1drv.ms/u/s!Av4d8txhBaH6htMDvZTE2MAZzQtynw)

**Comment by housseindjirdeh** [comment]
Thank you mate appreciate it :) @jacodv

**Comment by audrummer15** [comment]
This is awesome! I'm close with my build, but still getting the following error:

```
ERROR Error: Uncaught (in promise): SecurityError: Failed to execute 'replaceState' on 'History': A history state object with URL 'file:///home/user/dev/dist/' cannot be created in a document with origin 'null' and URL 'file:///home/user/dev/dist/index.html'
```

Has anyone ran into this, and overcome it? I'm at a loss. Thanks for the help so far!

**Comment by audrummer15** [comment]
In response to my above comment, this was failing due to the fact that we were using PushState routing. Changing over to the HashLocationStrategy made everything work! Still can't read the json files from disk though...

**Comment by jacodv** [comment]
I suggest you use a script tab to facilitate the loading of the JSON.  Ver reverted to var myvar = {//theJSON} to load it

**Comment by Justsahid** [comment]
Thanks @Markovy  @audrummer15 

I got it working completely in a fairly complex angular 2 app with multiple nested routes.

All I needed was

1. Router configuration with
`CommonModule,RouterModule.forRoot(routes**,{useHash:true}**) `

2. and this in the index.html file 
 Removed <base Href="/"> tag from html and added it like this.
` <script>document.write('<base href="' + document.location + '" />');</script>`

**Comment by AlwaysAbhl001** [comment]
@jacodv  unable to load my json files in i18n folders. ` dist\assets\i18n`

**Comment by jacodv** [comment]
How are you referencing them?

**Comment by ZebGirouard** [comment]
Oh my god, I cannot express my gratitude for this fix.  GH-pages for Angular has changed so much, and I tried sooooo many things before I found this thread.  @Markovy , you are my hero!  Thank you thank you!

**Comment by bentzi** [comment]
Thanks @Markovy

**Comment by jhonattan471** [comment]
Thanks @Justsahid save the day

**Comment by samuelkavin** [comment]
I got same problem here. router.navigate doesnt work in mobile browsers. Im looking for solution almost 1 day. Here is my code. I already tried all those examples in above. Here is my code

singup.ts
` onSubmit() {
    this.registrationMerchantService.registerInterestDirectMerchant({
      'email': this.personEmail
    }).subscribe( response => {
      this.router.navigate(['./submit-interest-success']);
    })
  }`

signup.routing.ts
```
`export const MODULE_ROUTES: Route[] = [
    { 
        path: '', 
        component: SignupMerchantComponent,
        children: [
            { path: 'submit-interest', component: SubmitInterestComponent },
            { path: 'submit-interest-success', component: SubmitInterestSuccessComponent },
            { path: '', redirectTo: 'submit-interest', pathMatch: 'full' }
        ] 
    }
]

export const appRouting = RouterModule.forRoot(MODULE_ROUTES);


export const SIGNUP_COMPONENTS = [
    SignupMerchantComponent,
    SignupSidebarComponent,

]`
```

signup.module.ts
```
`@NgModule({
  imports: [
    RouterModule.forChild(MODULE_ROUTES),
    SharedModule,
    CommonModule,
    FormsModule,
  ],
  declarations: [SIGNUP_COMPONENTS],
  providers: [
    SignupMerchantService,
  ]
})
export class SignupMerchantModule {
}`
```

@jacodv @markgoho @Koshmaar

**Comment by jhonattan471** [comment]
Hey, I fixed this a while ago,


If i remenber well the solution is in the index.html


let the <base href> in the html like this:


<!doctype html>
<html lang="en">

<head>
<script>document.write('<base href="' + document.location + '" />');</script>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="icon" type="image/x-icon" href="favicon.ico">
</head>

<body>
<app-root></app-root>
</body>

</html>


and in the rounter put this, ( add the usehash:true)

@NgModule({
declarations: [
AppComponent
],
imports: [
BrowserModule,
RouterModule.forRoot(appRoutes,{useHash:true}),
],
providers: [],
bootstrap: [AppComponent]
})
export class AppModule { }


________________________________
De: Samuel Kavin Loganathan <notifications@github.com>
Enviado: sexta-feira, 23 de fevereiro de 2018 03:54Para: angular/angular
Cc: Jhonattan Oliveira; Comment
Assunto: Re: [angular/angular] Cannot run angular 2+ from file:/// - looks like 'base href="/"' is the issue (#13948)


I got same problem here. router.navigate doesnt work in mobile browsers. Im looking for solution almost 1 day. Here is my code. I already tried all those examples in above. Here is my code

singup.ts
onSubmit() { this.registrationMerchantService.registerInterestDirectMerchant({ 'email': this.personEmail }).subscribe( response => { this.router.navigate(['./submit-interest-success']); }) }

signup.routing.ts

`export const MODULE_ROUTES: Route[] = [
    {
        path: '',
        component: SignupMerchantComponent,
        children: [
            { path: 'submit-interest', component: SubmitInterestComponent },
            { path: 'submit-interest-success', component: SubmitInterestSuccessComponent },
            { path: '', redirectTo: 'submit-interest', pathMatch: 'full' }
        ]
    }
]

export const appRouting = RouterModule.forRoot(MODULE_ROUTES);


export const SIGNUP_COMPONENTS = [
    SignupMerchantComponent,
    SignupSidebarComponent,

]`


signup.module.ts

`@NgModule({
  imports: [
    RouterModule.forChild(MODULE_ROUTES),
    SharedModule,
    CommonModule,
    FormsModule,
  ],
  declarations: [SIGNUP_COMPONENTS],
  providers: [
    SignupMerchantService,
  ]
})
export class SignupMerchantModule {
}`


@jacodv<https://github.com/jacodv> @markgoho<https://github.com/markgoho> @Koshmaar<https://github.com/koshmaar>

**Comment by mluis** [comment]
Hi, I'm getting errors alike the ones above. Not sure if this is the right thread to post though.

`ng build && electron .` build the app on the dist/ folder and launches electron with the index.html in the path.

Got electron loading the angular app but still having issues like:

`zone.js:2935 GET file:///assets/i18n/en.json net::ERR_FILE_NOT_FOUND`

Any idea how to overcome this step?

**Comment by jhonattan471** [comment]
Do you can send me your project to me see if I can fix it, or a example with the same error

Obter o Outlook para Android<https://aka.ms/ghei36>

________________________________

**Comment by mluis** [comment]
@jhonattan471 Thank you for the availability.

Just managed to solve the problem. The thing was ngx-translate was using an HTTP loader that didn't used the `./` . I managed to overcome the issue with:

```
export function createTranslateLoader(http: HttpClient) {
  return new TranslateHttpLoader(http, './assets/i18n/', '.json');
}
```

**Comment by rmorrise** [comment]
Is there any reason why this base-href script can't be included in all new projects generated by the CLI?

**Comment by Shekharrajak** [comment]
Thanks to this comment https://github.com/angular/angular/issues/13948#issuecomment-302727428 :+1:

**Comment by oneal1801** [comment]
Thank you so much for this fix @Markovy @Justsahid @jhonattan471 and @jacodv  for your post, you guys are the best!!

**Comment by leonardoperrone** [comment]
[#32 ](https://github.com/angular-schule/angular-cli-ghpages/issues/32#issuecomment-393556925) also helped me!

**Comment by davecar21** [comment]
> Thanks @Markovy @audrummer15
> 
> I got it working completely in a fairly complex angular 2 app with multiple nested routes.
> 
> All I needed was
> 
> 1. Router configuration with
>    `CommonModule,RouterModule.forRoot(routes,{useHash:true}) `
> 2. and this in the index.html file
>    Removed base Href="/" tag from html and added it like this.
>    ` <script>document.write('<base href="' + document.location + '" />');</script>`


I have tried this solution and its working fine.
But recently I encountered an issue when I add pictures on ` ./assets` folder and added a packages like ng-chartjs ..

![image](https://user-images.githubusercontent.com/17526518/51819848-e29d8880-230e-11e9-9963-68475c01e31b.png)


Any idea how to fix this?

**Comment by rostasan** [comment]
@davecar21 Did you find a solution to this? I just noticed this in my angular-cli build as well.
I have tsconfig that maps assets and other root folders, so i tried using the full ../../ path, but both methods result in the same error you're getting.

**Comment by davecar21** [comment]
@rostasan I got mine working by setting the `base-href` based on what URL you will deploy your app.
For example, since I will deploying my app in this URL www.example.com `ng build --prod --base-href=https://www.example.com`.

**Comment by qhuynh-bcs** [comment]
Is there any way to avoid add the hashtag (#) in the URL ?

**Comment by mlc-mlapis** [comment]
@qhuynh-bcs ... use `PathLocationStrategy` instead of `HashLocationStrategy` ... https://codecraft.tv/courses/angular/routing/routing-strategies

`RouterModule.forRoot(routes, { enableTracing: false, useHash: false });`

**Comment by qhuynh-bcs** [comment]
> Thanks @Markovy @audrummer15
> 
> I got it working completely in a fairly complex angular 2 app with multiple nested routes.
> 
> All I needed was
> 
> 1. Router configuration with
>    `CommonModule,RouterModule.forRoot(routes,{useHash:true}) `
> 2. and this in the index.html file
>    Removed base Href="/" tag from html and added it like this.
>    ` <script>document.write('<base href="' + document.location + '" />');</script>`

can fix this issue without useHash? Or is there any others way please ?
